### PR TITLE
Add option to throw on output toString and toJSON calls

### PR DESCRIPTION
**Pull request by devnev**
# Description

This adds an environment variable that changes output's `.toString` and `.toJSON` methods to throw an exception instead of returning a string containing a warning.

This feature mitigates the issue described in #4261 by allowing users to opt in to having accidental toString calls (e.g. in a JS backtick string that is missing `pulumi.interpolate`) throw an exception. The behaviour ends up similar to the `get()` method of outputs, which throws an exception.

I don't know how to write tests for this change, or even test it manually.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

File: sdk/nodejs/runtime/state.ts [hunk 1]
```diff
@@ -31,6 +31,8 @@ const nodeEnvKeys = {
     // this value is not set by the CLI and is controlled via a user set env var unlike the values above
     cacheDynamicProviders: "PULUMI_NODEJS_CACHE_DYNAMIC_PROVIDERS",
     organization: "PULUMI_NODEJS_ORGANIZATION",
+    // this value is not set by the CLI and is controlled via a user set env var unlike the values above
+    throwOutputToString: "PULUMI_NODEJS_THROW_OUTPUT_TOSTRING",
```
Comment (Frassle) [inline sdk/nodejs/runtime/state.ts:35]: I'd probably call this something like "PULUMI_ERROR_OUTPUT_STRING" as it would make sense to use the same envvar in other languages.

Further it would make sense to add this to sdk/go/common/env/env.go so it shows up in the CLI's env help.

File: sdk/nodejs/runtime/settings.ts [hunk 2]
```diff
@@ -533,3 +536,10 @@ export async function monitorSupportsDeletedWith(): Promise<boolean> {
 export async function monitorSupportsAliasSpecs(): Promise<boolean> {
     return monitorSupportsFeature("aliasSpecs");
 }
+
+/**
+ * Returns true if we throw exceptions when attempting to convert outputs to strings.
+ */
+export function throwOutputToString(): boolean {
+    return options().throwOutputToString ?? false;
```
Comment (Frassle) [inline sdk/nodejs/runtime/settings.ts:544]: Given this is just an envvar it should probably be in sdk/nodejs/utils.ts, not here and not on the settings object. See how disableResourceReferences is written.

File: sdk/nodejs/output.ts [hunk 3]
```diff
@@ -216,6 +219,9 @@ To get the value of an Output as a JSON value or JSON string consider either:
 
 See https://www.pulumi.com/docs/concepts/inputs-outputs for more details.
 This function may throw in a future version of @pulumi/pulumi.`;
+            if (settings.throwOutputToString()) {
+                throw new Error(message);
```
Comment (Frassle) [inline sdk/nodejs/output.ts:223]: Probably shouldn't include "is function may throw in a future version of @pulumi/pulumi." in the message thrown here.

File: sdk/nodejs/runtime/state.ts (hunk 1 above)
Comment (devnev) [reply sdk/nodejs/runtime/state.ts:35]: the Go SDK at least already has the JSON marshaler on outputs return an error, e.g. https://github.com/pulumi/pulumi/blob/05f64755af2c578c6fb41a4c1427063b058123f4/sdk/go/pulumi/types.go#L257, so this option may not apply to all languages

File: sdk/nodejs/runtime/settings.ts (hunk 2 above)
Comment (devnev) [reply sdk/nodejs/runtime/settings.ts:544]: Interesting, I was basing this off of `cacheDynamicProviders` / `PULUMI_NODEJS_CACHE_DYNAMIC_PROVIDERS` which is also only an env var. Seems simpler to do int the `disableResourceReferences`, but on the other hand that is adding to a generic `utils.ts` module instead of a more specific `settings.ts` module.

File: sdk/nodejs/runtime/state.ts (hunk 1 above)
Comment (devnev) [reply sdk/nodejs/runtime/state.ts:35]: The python SDK on the other hand also returns a string instead of raising an error:
https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/output.py#L725

**Comment by devnev** [comment]
Thanks for taking a look at the PR!

File: sdk/nodejs/runtime/state.ts (hunk 1 above)
Comment (Frassle) [reply sdk/nodejs/runtime/state.ts:35]: > the Go SDK at least already has the JSON marshaler on outputs return an error, e.g.

That's specific to JSON marshaling. Go doesn't let you call `String()` or `sprintf("%s")` on outputs.

> The python SDK on the other hand also returns a string instead of raising an error:

Yes, same as the node SDK, and the Java and dotnet sdk.

> so this option may not apply to all languages

It won't apply to _all_ languages but it will apply to _most_ languages. It should be named generally, not specific to NODEJS.

File: sdk/nodejs/runtime/settings.ts (hunk 2 above)
Comment (Frassle) [reply sdk/nodejs/runtime/settings.ts:544]: `cacheDynamicProviders` is being needlessly complex in this case, do the same as `disableResourceReferences`. I'll change `cacheDynamicProviders` to this style as well when I've got 10 minutes.

File: sdk/nodejs/runtime/state.ts (hunk 1 above)
Comment (devnev) [reply sdk/nodejs/runtime/state.ts]: Thanks for the clarification. I've added a declaration to `sdk/go/common/env/env.go`, should I be doing anything else with that variable?

**Comment by Frassle** [comment]
> should I be applying this to other SDKs, or is doing only the NodeJS SDK fine for now?

It would be nice if you did the others (well only python needs doing in this repo) but I'll accept this just changing node and I'll go fix the others up afterwards.

File: sdk/nodejs/runtime/state.ts (hunk 1 above)
Comment (Frassle) [reply sdk/nodejs/runtime/state.ts]: No, just need it there so it can show up in env help.

File: changelog/pending/20231210--sdk-nodejs--add-option-to-throw-on-output-tostring-and-tojson-calls.yaml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: sdk/nodejs
+  description: Add option to throw on output toString and toJSON calls
```
Comment (Frassle) [inline changelog/pending/20231210--sdk-nodejs--add-option-to-throw-on-output-tostring-and-tojson-calls.yaml:4]:
```suggestion
  description: Add an environment variable (`PULUMI_ERROR_OUTPUT_STRING`) to throw when an output tries to convert to a string.
```

File: sdk/nodejs/tests/sxs_ts_test/yarn.lock [hunk 5]
```diff
@@ -1209,6 +1209,11 @@ typed-array-length@^1.0.4:
     for-each "^0.3.3"
     is-typed-array "^1.1.9"
 
+typescript@^4:
```
Comment (Frassle) [inline sdk/nodejs/tests/sxs_ts_test/yarn.lock:1212]: Can you revert this change. It's a bug in the test runner I need to fix up.

**Comment by devnev** [comment]
woops, forgot to run make format & lint on the python sdk after the additions. fixed now.

**Review by Frassle — APPROVED** [review]

**Comment by Frassle** [comment]
Thanks for this. I'll take a look at updating dotnet and java SDKs to use this envvar as well.
